fedimint-client/src/lib.rs

```diff
@@ -317,20 +323,88 @@ pub type ModuleGlobalContextGen = ContextGen<DynGlobalClientContext>;
 impl Client {
     /// Add funding and/or change to the transaction builder as needed, finalize
     /// the transaction and submit it to the federation.
-    pub async fn finalize_and_submit_transaction(
+    pub async fn finalize_and_submit_transaction<F, M>(
         &self,
         operation_id: OperationId,
+        operation_type: &str,
+        operation_meta: F,
         tx_builder: TransactionBuilder,
-    ) -> anyhow::Result<TransactionId> {
+    ) -> anyhow::Result<TransactionId>
+    where
+        F: Fn(TransactionId) -> M,
+        M: serde::Serialize,
+    {
         let mut dbtx = self.inner.db.begin_transaction().await;
+
+        if ClientInner::operation_exists(&mut dbtx, operation_id).await {
+            bail!("There already exists an operation with id {operation_id:?}")
+        }
+
         let txid = self
             .inner
             .finalize_and_submit_transaction(&mut dbtx, operation_id, tx_builder)
             .await?;
+
+        dbtx.insert_entry(
```

Comment:
Use `insert_new_entry` if you don't want to overwrite it. Not sure if it's worth keeping `operation_exists` given that. Might be more human readable, but then it's an extra codepath...

Reply:
Good catch! The previous check has to stay in place though since all the `new_` version does is *warn* in the logs if there was a replacement happened.

It might be worth making it an assertion, back in the day when `insert_new`/`delete_existing` were introduced it was really hard to know at times if there was no case in which a key would exist since there weren't any database transactions yet and merely these DB change batches that would be executed at a later point. By now we might be reasonably sure what we expect and could even return `Option<V>` from `insert` and then check at the call site if the returned element is what we expected.

Reply:
Created #2321